[ONNX] Support torch.device in FX exporter

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105757

Fixes https://github.com/pytorch/pytorch/issues/105172

When torch.device appears in kwargs, we ignore it, as it's not needed in onnx. However, if it's in args, it's used by dispatcher, and we didn't handle it. This PR adds torch.device into args processing.

titaiwangms committed:
[ONNX] Support torch.device in FX exporter

[ghstack-poisoned]

titaiwangms commented:
Is there an easy test that I could use? I found that prims are not yet tested in op_test.py.

thiagocrepaldi commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -184,6 +183,10 @@ def filter_incompatible_and_dtype_convert_kwargs(kwargs):
                 value = int(jit_type_utils.JitScalarType.from_dtype(value).onnx_type())
+        if key == "device":
+            # torch.device is not supported by onnxscript (no op). We turn it into
+            # a string.
```
how will the `onnx_kwargs = {"device": str(value)}` be used later on when we call `        symbolic_fn = onnxfunction_dispatcher.dispatch(node=node, onnx_args=onnx_args, onnx_kwargs=onnx_kwargs, diagnostic_context=self.diagnostic_context,)`? Will it be just ignored or it will be part of the variant matching through attributes? That might be an useful information for the dostring of `filter_incompatible_and_dtype_convert_kwargs` and the PR description

thiagocrepaldi reviewed (commented):
LGTM. Don't forget to add a PR description and unit test to make sure the device information is properly processed

titaiwangms committed:
Update on "[ONNX] Support torch.device in FX exporter"


Fixes https://github.com/pytorch/pytorch/issues/105172

[ghstack-poisoned]

justinchuby commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -636,6 +636,24 @@ def test_gpt2_tiny(self):
+            def forward(self, x):
+                # Assuming x is a tensor on the CPU, move it to the desired device using device_put()
+                x = torch.ops.prims.device_put(
+                    x, "cuda" if torch.cuda.is_available() else "cpu"
```
Will this break the graph?

titaiwangms replied:
Not sure, I will update it to only cpu

titaiwangms committed:
Update on "[ONNX] Support torch.device in FX exporter"


Fixes https://github.com/pytorch/pytorch/issues/105172

When torch.device appears in kwargs, we ignore it, as it's not needed in onnx. However, if it's in args, it's used by dispatcher, and we didn't handle it. This PR adds torch.device into args processing. 

[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -636,6 +636,22 @@ def test_gpt2_tiny(self):
             model, [], inputs, additional_test_inputs=[((), another_inputs)]
         )
 
+    @pytorch_test_common.skip_min_ort_version(
```
How long should we keep those? I feel we can delete but like to hear your thoughts @justinchuby @titaiwangms

titaiwangms replied:
TBH, I think we are free to delete it. But I am not sure if there is any internal tests using older version of ort.....

BowenBao replied:
haha there shouldn't, let's do it separately then.

titaiwangms commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -636,6 +636,22 @@ def test_gpt2_tiny(self):
+        class CustomModule(nn.Module):
+            def forward(self, x):
+                # Assuming x is a tensor on the CPU, move it to the desired device using device_put()
+                x = torch.ops.prims.device_put(x, "cpu")
```
Need to update onnx-script version

titaiwangms committed:
Update on "[ONNX] Support torch.device in FX exporter"


Fixes https://github.com/pytorch/pytorch/issues/105172

When torch.device appears in kwargs, we ignore it, as it's not needed in onnx. However, if it's in args, it's used by dispatcher, and we didn't handle it. This PR adds torch.device into args processing. 

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge

titaiwangms committed:
Update on "[ONNX] Support torch.device in FX exporter"


Fixes https://github.com/pytorch/pytorch/issues/105172

When torch.device appears in kwargs, we ignore it, as it's not needed in onnx. However, if it's in args, it's used by dispatcher, and we didn't handle it. This PR adds torch.device into args processing. 

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
